index.js

```diff
@@ -25,112 +25,116 @@ class Turtler {
-        if(!hasHeader || !headerSeparator) return;
+        table += `${row}\n`;
 
-        // we add columnSeparator.length because above we joined on that string which could be more than one character.
-        // We only want one character so we only look at the first character of the string
-        table += headerSeparator[0].repeat((columnWidths.reduce((a, b) => a + b)) + columnSeparator.length) + '\n';
-      }
-    });
+        if(l === 0) {
+          // ignore the header string if hasHeader is false or headerSeparator is not set
+          if(!hasHeader || !headerSeparator) return;
 
-    return table;
+          // we add columnSeparator.length because above we joined on that string which could be more than one character.
+          // We only want one character so we only look at the first character of the string
+          table += headerSeparator[0].repeat((columnWidths.reduce((a, b) => a + b)) + columnSeparator.length) + '\n';
+        }
+      });
+      this.asciiTable = table;
+    }
+    return this.asciiTable;
   }
   /**
    * renders a markdown table
    * @method markdown
    * @return {String} markdown table string
    */
   markdown() {
-    const { data } = this;
-
-    let table = '';
-    let columns = 0;
-    let columnWidths = [];
-
-    // Find the maximum width of each column
-    // If rows contain uneven number of columns, throw
-    data.forEach((row) => {
-      // The row should be an array
-      if(!Array.isArray(row)) throw new Error('data should be an array of arrays');
-      // Set the initial length of the row
-      if(!columns) columns = row.length;
-      // If the current row is not the same length as the initial one throw error
-      if(columns !== row.length) throw new Error('columns are not formed properly');
-
-      // find the maximum length of each column
-      row.forEach((v, l) => {
-        // column values must be strings
-        if(typeof v !== 'string') throw new Error('column values should be strings');
-
-        // Find the maximum string length in each column
-        if(!columnWidths[l] || columnWidths[l] < v.length) {
-          columnWidths[l] = v.length;
-        }
-      });
-    });
-
-    // make the rows nice and tidy giving enough space on all sides to make it uniform
-    data.forEach((row, l) => {
-      row = row.map((value, i) => {
-        // Create pad of empty spaces to match the width of this value to max width of this column
-        return value + ' '.repeat(columnWidths[i] - value.length);
-      // join on the pipe which denotes the border of a table element in markdown
-      }).join(' | ');
-
-      table += `| ${row} |\n`;
-
-      if(l === 0) {
-        // we add the header with the width of the column minus the addition of a pipe symbol
-        table += columnWidths.map((width, i) => {
-          if(i == 0) {
-            return '|' + '-'.repeat(width + 2) + '|';
+    if (!this.markdownTable) {
```

Comment:
why not make this a 

```js
if(this.markdownTable) return this.markdownTable
```

It is a smaller change and doesn't make the code hard to decipher.